File: leet_1424_diagonal_traverse_ii.cc
class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& nums)
    {
        std::map<int, vector<int>> D;
        int r = -1, c;
        while (++r < nums.size())
        {
            c = -1;
            while (++c < nums[r].size())
                D[r + c].push_back(nums[r][c]);
        }
        vector<int> res;
        auto p = D.begin();
        while (p != D.end())
        {
            while ( ! p->second.empty())
            {
                res.push_back(p->second.back());
                p->second.pop_back();
            }
            ++p;
        }
        return res;
    }
};

File: leet_1424_diagonal_traverse_ii.ts
function findDiagonalOrder(nums: number[][]): number[] {
    let D: Record<number, number[]> = []
    let r = -1, c
    while (++r < nums.length) {
        c = -1
        while (++c < nums[r].length) {
            if ( ! D.hasOwnProperty(r + c)) {
                D[r + c] = [nums[r][c]]
            } else {
                D[r + c].push(nums[r][c])
            }
        }
    }
    console.log(D)
    let res: number[] = []
    for (let [_, arr] of Object.entries(D)) {
        while (arr.length) {
            res.push(arr.pop())
        }
    }
    return res
};
